openvasd/openvasd.c

```diff
@@ -0,0 +1,2251 @@
+  if (response == NULL)
+    return NULL;
+
+  init_openvasd_stringstream (&resp);
+  if ((hnd = handler (conn, POST, "/scans", data, &resp, &err)) == NULL)
+    {
+      g_free (resp.ptr);
+      response->code = RESP_CODE_ERR;
+      response->body = err;
+      return response;
+    }
+
+  openvasd_send_request (hnd, NULL, response);
+  if (response != NULL && response->code != RESP_CODE_ERR)
+    response->body = g_strdup (resp.ptr);
+  else if (response == NULL || response->code == RESP_CODE_ERR)
+    {
+      g_free (resp.ptr);
+      response->code = RESP_CODE_ERR;
+      if (response->body == NULL)
+        response->body =
+          g_strdup ("{\"error\": \"Storing scan configuration\"}");
+      g_warning ("%s: Error storing scan configuration ", __func__);
+      return response;
+    }
+
+  // Get the Scan ID
+  parser = cJSON_Parse (response->body);
+  if (!parser)
+    {
+      const char *error_ptr = cJSON_GetErrorPtr ();
+      if (error_ptr != NULL)
+        {
+          response->body = g_strdup_printf ("{\"error\": \"%s\"}", error_ptr);
+          g_warning ("%s: %s", __func__, error_ptr);
+        }
+      else
+        {
+          response->body = g_strdup (
+            "{\"error\": \"Parsing json string to get the scan ID\"}");
+          g_warning ("%s: Parsing json string to get the scan ID", __func__);
+        }
+      response->code = RESP_CODE_ERR;
+      g_free (resp.ptr);
+      cJSON_Delete (parser);
+      return response;
+    }
+
+  (*conn)->scan_id = g_strdup (cJSON_GetStringValue (parser));
+
+  // Start the scan
+  path = g_string_new ("/scans");
+  if ((*conn)->scan_id != NULL && (*conn)->scan_id[0] != '\0')
+    {
+      g_string_append (path, "/");
+      g_string_append (path, (*conn)->scan_id);
+    }
+  else
+    {
+      response->code = RESP_CODE_ERR;
+      response->body = g_strdup ("{\"error\": \"Missing scan ID\"}");
+      g_string_free (path, TRUE);
+      g_warning ("%s: Missing scan ID", __func__);
+      cJSON_Delete (parser);
+      return response;
+    }
+
+  reset_openvasd_stringstream (&resp);
+  if ((hnd = handler (conn, POST, path->str, "{\"action\": \"start\"}", &resp,
+                      &err))
+      == NULL)
+    {
+      g_free (resp.ptr);
+      g_string_free (path, TRUE);
+      response->code = RESP_CODE_ERR;
+      response->body = err;
+      return response;
+    }
+  g_string_free (path, TRUE);
+
+  openvasd_send_request (hnd, NULL, response);
+  if (response != NULL && response->code != RESP_CODE_ERR)
+    response->body = g_strdup (resp.ptr);
+  else if (response == NULL || response->code == RESP_CODE_ERR)
+    {
+      g_free (resp.ptr);
+      response->code = RESP_CODE_ERR;
+      if (response->body == NULL)
+        response->body = g_strdup ("{\"error\": \"Starting the scan.\"}");
+      g_warning ("%s: Error starting the scan.", __func__);
+      return response;
+    }
+
+  cJSON_Delete (parser);
+  response->body = g_strdup (resp.ptr);
```

Comment:
Maybe I'm missing some logic here, but it looks like body is assigned above (eg at line 779), and then assigned again here, leaking the first assign.

Also in the error cases body is assigned even though it had already been assigned.

Maybe would be cleaner to allocate the responses at the points you're returning them.